tensorflow/java/src/gen/cc/java_defs.h

```diff
@@ -0,0 +1,266 @@
+
+ private:
+  string attrs_;
+
+  Annotation(const string& name, const string& package)
+    : Type(Kind::ANNOTATION, name, package) {}
+};
+
+/// \brief A definition of a Java variable
+///
+/// This class defines an instance of a type, which could be documented.
+class Variable {
+ public:
+  static Variable Of(const string& name, const Type& type) {
+    return Variable(name, type, false);
+  }
+  static Variable VarArg(const string& name, const Type& type) {
+    return Variable(name, type, true);
+  }
+  const string& name() const { return name_; }
+  const Type& type() const { return type_; }
+  bool variadic() const { return variadic_; }
+  const string& descr() const { return descr_; }
+  Variable& descr(const string& descr) { descr_ = descr; return *this; }
+
+ private:
+  string name_;
+  Type type_;
+  bool variadic_;
+  string descr_;
+
+  Variable(const string& name, const Type& type, bool variadic)
+    : name_(name), type_(type), variadic_(variadic) {}
+};
+
+/// \brief A definition of a Java class method
+///
+/// This class defines the signature of a method, including its name, return
+/// type and arguments.
+class Method {
+ public:
+  static Method Of(const string& name, const Type& return_type) {
+    return Method(name, return_type, false);
+  }
+  static Method ConstructorFor(const Type& clazz) {
+    return Method(clazz.name(), clazz, true);
+  }
+  bool constructor() const { return constructor_; }
+  const string& name() const { return name_; }
+  const Type& ret_type() const { return ret_type_; }
+  const string& descr() const { return descr_; }
+  Method& descr(const string& descr) { descr_ = descr; return *this; }
+  const string& ret_descr() const { return ret_descr_; }
+  Method& ret_descr(const string& descr) { ret_descr_ = descr; return *this; }
+  const std::vector<Variable>& args() const { return args_; }
+  Method& args(const std::vector<Variable>& args) {
+    args_.insert(args_.cend(), args.cbegin(), args.cend());
+    return *this;
+  }
+  Method& arg(const Variable& var) { args_.push_back(var); return *this; }
+  const std::vector<Annotation>& annotations() const { return annotations_; }
+  Method& annotation(const Annotation& annotation) {
+    annotations_.push_back(annotation);
+    return *this;
+  }
+
+ private:
+  string name_;
+  Type ret_type_;
+  bool constructor_;
+  string descr_;
+  string ret_descr_;
+  std::vector<Variable> args_;
+  std::vector<Annotation> annotations_;
+
+  Method(const string& name, const Type& ret_type, bool constructor)
+    : name_(name), ret_type_(ret_type), constructor_(constructor) {}
+};
+
+/// \brief A piece of code to read from a file.
+class Snippet {
+ public:
+  explicit Snippet(const string& fname, Env* env = Env::Default()) {
+    TF_CHECK_OK(ReadFileToString(env, fname, &data_));
+  }
+  const string& data() const { return data_; }
+
+ private:
+  string data_;
+};
+
+// Templates implementation
```

Comment:
I'd like to see if we can avoid templates. Perhaps remove this from this PR for now and we can discuss options on the follow up PR that will need to introduce some sort of scanner?

Reply:
Ok no problem

Reply:
Ok no problem